### PR TITLE
Add support for multiple flexattention calls in a single compile

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125516

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -179,11 +179,12 @@ def trace_flex_attention(
     ] + [torch.zeros((), dtype=torch.int) for _ in range(4)]
     with TransformGetItemToIndex():
         score_graph = make_fx(score_mod)(*example_vals, *other_buffers)
-    proxy_mode.tracer.root.register_module("sdpa_score", score_graph)
+    qualname = proxy_mode.tracer.get_fresh_qualname("sdpa_score")
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:182]: would we hard error before or just overrwrite?

File: torch/_higher_order_ops/flex_attention.py [hunk 2]
```diff
@@ -179,11 +179,12 @@ def trace_flex_attention(
...
+    proxy_mode.tracer.root.register_module(qualname, score_graph)
     node_args = (query, key, value, score_graph, *other_buffers)
     proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
     out_proxy = proxy_mode.tracer.create_proxy(
-        "call_function", flex_attention, proxy_args, {}, name="flex_attention"
+        "call_function", flex_attention, proxy_args, {}
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:187]: do we loos nice names?

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -3587,8 +3587,8 @@ def __init__(
         self.mutated_inputs = mutated_inputs
         if mutated_inputs is not None:
             # Ensure that the mutated inputs are only allowed for certain nodes
-            allowed_set = {"flex_attention"}
-            current_node = str(V.graph.current_node)
+            allowed_set = {torch.ops.higher_order.flex_attention}
```
Comment (drisspg) [inline torch/_inductor/ir.py:3590]: this is better

File: torch/fx/_symbolic_trace.py [hunk 4]
```diff
@@ -309,6 +309,20 @@ def __init__(
         # Mapping of node name to module scope
         self.node_name_to_scope: Dict[str, Tuple[str, type]] = {}
 
+    @compatibility(is_backward_compatible=True)
+    def get_fresh_qualname(self, prefix: str) -> str:
```
Comment (drisspg) [inline torch/fx/_symbolic_trace.py:313]: what if you store a dict with the seen prefixes and then just bump the count if the prefix is in the dict? 

maybe premature optimization but could this while loop get pretty long in practice?

File: torch/_higher_order_ops/flex_attention.py (hunk 1 above)
Comment (Chillee) [reply torch/_higher_order_ops/flex_attention.py:182]: silently incorrect.

File: torch/_higher_order_ops/flex_attention.py (hunk 2 above)
Comment (Chillee) [reply torch/_higher_order_ops/flex_attention.py:187]: No the name is still based off the hop operator.

File: torch/fx/_symbolic_trace.py (hunk 4 above)
Comment (Chillee) [reply torch/fx/_symbolic_trace.py:313]: yeah maybe. But all the existing examples just did this.

Comment (ezyang) [reply torch/fx/_symbolic_trace.py:313]: We actually fixed a material compile time perf problem from exactly this pattern a while back, I can't find the diff, but please just fix the O(n^2) to not be n^2. A two line improvement is to maintain a counter and initialize i off of that.

Comment (Chillee) [reply torch/fx/_symbolic_trace.py:315]: ok fine I updated it :)

Comment (Chillee) [reply torch/fx/_symbolic_trace.py:315]: Unfortunately, this is causing a lot of annoyances for the thing I was worried about that led to me not adding it initially - it seems that in some cases the tracer state is persisted across different modules, causing a bunch of churn in various tests 🤔

**Review by yanboliang — APPROVED** [review]

**Review by drisspg — APPROVED** [review]

**Pull request closed without merging**